src/style.js

```diff
@@ -1,53 +1,30 @@
-import { Component } from 'react'
+import { useLayoutEffect } from 'react'
 import StyleSheetRegistry from './stylesheet-registry'
 
 const styleSheetRegistry = new StyleSheetRegistry()
 
-export default class JSXStyle extends Component {
-  constructor(props) {
-    super(props)
-    this.prevProps = {}
-  }
-
-  static dynamic(info) {
-    return info
-      .map(tagInfo => {
-        const baseId = tagInfo[0]
-        const props = tagInfo[1]
-        return styleSheetRegistry.computeId(baseId, props)
-      })
-      .join(' ')
-  }
-
-  // probably faster than PureComponent (shallowEqual)
-  shouldComponentUpdate(otherProps) {
-    return (
-      this.props.id !== otherProps.id ||
-      // We do this check because `dynamic` is an array of strings or undefined.
-      // These are the computed values for dynamic styles.
-      String(this.props.dynamic) !== String(otherProps.dynamic)
-    )
-  }
-
-  componentWillUnmount() {
-    styleSheetRegistry.remove(this.props)
+export default function JSXStyle(props) {
+  if (typeof window === 'undefined') {
+    styleSheetRegistry.add(props)
+    return null
   }
-
-  render() {
-    // This is a workaround to make the side effect async safe in the "render" phase.
-    // See https://github.com/zeit/styled-jsx/pull/484
-    if (this.shouldComponentUpdate(this.prevProps)) {
-      // Updates
-      if (this.prevProps.id) {
-        styleSheetRegistry.remove(this.prevProps)
-      }
-
-      styleSheetRegistry.add(this.props)
-      this.prevProps = this.props
+  useLayoutEffect(() => {
+    styleSheetRegistry.add(props)
+    return () => {
+      styleSheetRegistry.remove(props)
     }
+  }, [props.id, String(props.dynamic)])
```

Comment:
id` is an hash of `children` so it might be fine to omit children but for correctness `children` should be in the deps array because it is a dependency too.

Reply:
observed some interesting transpile behavior only on **production** here:
`children` has been transformed to strings concatenation  🤔 then `JSXStyle` always get children as array so it will keep triggering hooks

```js
o(
  r.default, {
    id: '200133253',
    dynamic: [s, i],
  },
  [
    '.sth.__jsx-style-dynamic-selector{color:'
    .concat(s, ';background:')
    .concat(
      i,
      ';display:inline-block;font-weight:',
    )
    .concat(u ? 'bold' : 'normal', ';padding:')
  ]
)
```

Reply:
Yes it looks like a template literal with expressions, weird that it is evaluated and inlined. Anyway it should be ok to remove children from the deps!